examples/digital_fingerprinting/production/morpheus/dfp/utils/model_cache.py

```diff
@@ -0,0 +1,324 @@
+                        self._reg_model_name))
+
+                return model_cache
+        except TimeoutError:
+            logger.error("Deadlock detected while loading model cache. Please report this to the developers.")
+            raise RuntimeError("Deadlock detected while loading model cache")
+
+
+class ModelManager:
+
+    def __init__(self, model_name_formatter: str) -> None:
+        self._model_name_formatter = model_name_formatter
+
+        self._user_model_cache: typing.Dict[str, UserModelMap] = {}
+
+        self._model_cache: typing.Dict[str, ModelCache] = {}
+        self._model_cache_size_max = 100
+
+        self._cache_timeout_sec = 600
+
+        self._user_model_cache_lock = threading.RLock()
+        self._model_cache_lock = threading.RLock()
+
+        self._existing_models: typing.Set[str] = set()
+        self._existing_models_updated = datetime(1970, 1, 1)
+
+        # Force an update of the existing models
+        self._model_exists("")
+
+    @property
+    def cache_timeout_sec(self):
+        return self._cache_timeout_sec
+
+    def _model_exists(self, reg_model_name: str) -> bool:
+
+        now = datetime.now()
+
+        # See if the list of models needs to be updated
+        if ((now - self._existing_models_updated).seconds > self._cache_timeout_sec):
+
+            try:
+                with timed_acquire(self._model_cache_lock, timeout=1.0):
+
+                    logger.debug("Updating list of available models...")
+                    client = MlflowClient()
+
+                    results: PagedList[RegisteredModel] = PagedList([], token=None)
+
+                    # clear the set to hanfle the case where a model has been removed
+                    self._existing_models.clear()
+
+                    # Loop over the registered models with the pagination
+                    while ((results := client.list_registered_models(max_results=1000, page_token=results.token))
+                           is not None):
+
+                        self._existing_models.update(model.name for model in results)
+
+                        if (len(results.token) == 0):
+                            break
+
+                    logger.debug("Updating list of available models... Done.")
+
+                    # Save the update time
+                    self._existing_models_updated = now
+
+            except TimeoutError:
+                logger.error("Deadlock detected checking for new models. Please report this to the developers.")
+                raise RuntimeError("Deadlock detected checking for new models")
+            except Exception:
+                logger.exception("Exception occurred when querying the list of available models", exc_info=True)
+                raise
+
+        return reg_model_name in self._existing_models
+
+    def user_id_to_model(self, user_id: str):
+        return self._model_name_formatter.format(user_id=user_id)
```

Comment:
In the DFPMLFlowModelWriterStage the `user_md5` field is available, and should be presented here for parity. 

Probably best if they both share the same format function to ensure they remain in sync.

Reply:
Done